CB-84: Add checkbox to filter releases with reviews on artist discography pages.

ferbncode opened the pull request:
Added checkbox on artist discography pages to filter only those releases (on that page) which have reviews. 

To fetch only those releases which have a review (rather than filtering on that page), I guess we can later fetch all releases (rather than applying the limit) and then show them (shouldn't slow down the page).

ferbncode commented:
Some screenshots:
![Unchecked](https://user-images.githubusercontent.com/12391372/29587963-57f60354-87ad-11e7-9ea6-930383b86e41.png)
![Checked for showing only reviewed entities](https://user-images.githubusercontent.com/12391372/29587961-57f04f90-87ad-11e7-9b5b-eb0e48a2cbaf.png)
![No reviewed releases](https://user-images.githubusercontent.com/12391372/29587962-57f0f2d8-87ad-11e7-8bd0-b9b730050966.png)

alastair commented on critiquebrainz/frontend/templates/artist/entity.html:
```diff
@@ -48,7 +48,17 @@
+      <div class="col-sm-4">
+        <div>
+          <input type="checkbox" name="only-reviewed" value="1" class="only-reviewed" onchange="showOnlyReviewed()"/>
+          {{ _('Filter releases here with reviews') }}<br/>
```
"Only show releases with reviews"

alastair commented on critiquebrainz/frontend/templates/artist/entity.html:
```diff
@@ -66,13 +76,16 @@
         <a href="{{ url_for('artist.entity', mbid=artist.id) }}?release_type=other">{{ _('Other releases') }}</a>
       </li>
     </ul>
+    <div class="no_reviewed_entities">
```
if the default is to not check the box, can you make this hidden by default instead of hiding it in js?

alastair commented on critiquebrainz/frontend/templates/artist/entity.html:
```diff
@@ -48,7 +48,17 @@
+      </div>
+      <div class="col-sm-4">
+        <div>
+          <input type="checkbox" name="only-reviewed" value="1" class="only-reviewed" onchange="showOnlyReviewed()"/>
```
instead of using an html attribute here, other places in the CB code us jquery to install an event handler: https://github.com/metabrainz/critiquebrainz/blob/483594644b4954c3a62f89c84d9abf1e97b14a8a/critiquebrainz/frontend/templates/review/modify/base.html#L76
We should do the same thing for consistency

alastair commented on critiquebrainz/frontend/templates/artist/entity.html:
```diff
@@ -181,3 +194,24 @@
+{% block scripts %}
+  {{ super() }}
+  <script>
+  $(".no_reviewed_entities").hide();
```
With javascript it's a good idea to run commands only after the full page has loaded. You can use document.ready here: https://github.com/metabrainz/critiquebrainz/blob/483594644b4954c3a62f89c84d9abf1e97b14a8a/critiquebrainz/frontend/templates/review/modify/base.html#L63

gentlecat commented on critiquebrainz/frontend/templates/artist/entity.html:
```diff
@@ -48,7 +48,17 @@
+      <div class="col-sm-4">
+        <div>
+          <input type="checkbox" name="only-reviewed" value="1" id="only-reviewed"/>
+          {{ _('Only show releases with reviews') }}<br/>
```
Artist page doesn't have releases. Only release groups.

gentlecat commented on critiquebrainz/frontend/templates/artist/entity.html:
```diff
@@ -181,3 +194,25 @@
+        }
+      })
+    })
+  </script>
```
It's better to put this JS code into a separate script in *static/scripts*, build and import it.

gentlecat commented:
Also please don't forget to assign yourself to the issue in JIRA and mark it as WIP so that there's no confusion if someone else decides to work on it.

gentlecat commented:
On the last screenshot there's a "Next" button. Is that a bug or the screenshot is just a prototype?

ferbncode commented:
> On the last screenshot there's a "Next" button. Is that a bug or the screenshot is just a prototype?

The "Next" Button appeared as the filtering was being done on just that page. So, as there were no more reviewed entities (on that page), the "Next Button" appears. I've fixed this and now all reviewed release groups are shown together.

I've rebased the present branch and based it on PR #140. I've modified the code to fix the problem by fetching all release groups of an artist by directly using the db. 
Now all reviewed release groups are shown once the filter is checked. Also, the `Previous` and the `Next` buttons don't reload the entire page, but they just refresh the entities to be shown.

gentlecat commented:
Was trying to test this and ran into the following issue:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/werkzeug/wsgi.py", line 660, in __call__
    return app(environ, start_response)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/code/critiquebrainz/frontend/views/release_group.py", line 28, in entity
    release = mb_release.get_release_by_id(release_group['release-list'][0]['id'])
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/release.py", line 23, in get_release_by_id
    release = _get_release_by_id(mbid)
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/release.py", line 31, in _get_release_by_id
    includes=['media', 'release-groups'],
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/release.py", line 82, in fetch_multiple_releases
    releases = {str(mbid): to_dict_releases(releases[mbid], includes_data[releases[mbid].id]) for mbid in mbids}
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/release.py", line 82, in <dictcomp>
    releases = {str(mbid): to_dict_releases(releases[mbid], includes_data[releases[mbid].id]) for mbid in mbids}
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/serialize.py", line 193, in to_dict_releases
    for medium in includes['media']]
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/serialize.py", line 193, in <listcomp>
    for medium in includes['media']]
  File "/code/critiquebrainz/frontend/external/musicbrainz_db/serialize.py", line 153, in to_dict_medium
    'format': medium.format.name,
AttributeError: 'NoneType' object has no attribute 'name'
```

gentlecat commented:
I'm seeing this on `master` as well.

ferbncode commented:
> Even if most artists don't have many, some pages will probably be slowed down significantly the more release groups they have.

That is correct. Some pages slow down as they have many release groups (Various Artists page (no. of release groups: 145133), for instance). But that is important if all reviewed release groups are to be shown together. Some possible (though probably not the best) work-arounds can be:
+ adding a loader for making the user wait for the page for load.
+ keeping pagination and filtering the reviewed entities only on that very page.

I am unsure if either is a good choice. Please suggest.

paramsingh commented:
@ferbncode, I'd like to discuss how exactly to move forward on this, once you get the time.

paramsingh commented:
Closing this for now so that discussion can be moved to the ticket: https://tickets.metabrainz.org/browse/CB-84?focusedCommentId=45384&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-45384
